File: vector.js
/* 
  Classe Vector 
  Exemples de vectors: forces
  - no té dependències
*/

class Vector {
  constructor(x,y,z){
    this.x=x||0;//component x
    this.y=y||0;//component y
    this.z=z||0;//component z
  };

  //suma i resta 2 vectors
  suma (vec){return new Vector(this.x+vec.x, this.y+vec.y, this.z+vec.z);} //vector + vector = vector
  resta(vec){return new Vector(this.x-vec.x, this.y-vec.y, this.z-vec.z);} //vector - vector = vector

  //multiplicació per un escalar
  //divisió es pot fer fent this.multiplica(1/esc);
  multiplica(esc){return new Vector(this.x*esc, this.y*esc, this.z*esc);} //vector * escalar  = vector

  //producte escalar i vectorial entre 2 vectors
  prod_esc(vec){ return (vec.x*this.x + vec.y*this.y + vec.z*this.z);}                                                //vector · vector = escalar
  prod_vec(vec){ return new Vector(this.y*vec.z-this.z*vec.y, this.z*vec.x-this.x*vec.z, this.x*vec.y-this.y*vec.x);} //vector x vector = vector

  //modul i norma
  get modul(){ return Math.sqrt(this.x*this.x+this.y*this.y+this.z*this.z);} //llargada/magnitud d'un vector
  get norma(){ return this.multiplica(1/this.modul);}                            //normalitza vector per la seva llargada
}

module.exports=Vector;

File: structs/vector.c
#include "vector.h"

//mostra vector
void print_vector(Vector v){
  //mostra components i lloc memòria
  printf("Vector{x:%lf, y:%lf, z:%lf} @%p\n",
    v.x, v.y, v.z, v);
}

//vector + vector = vector
Vector suma(Vector v1, Vector v2){
  Vector v;
  v.x = v1.x + v2.x;
  v.y = v1.y + v2.y;
  v.z = v1.z + v2.z;
  return v;
}

//vector - vector = vector
Vector resta(Vector v1, Vector v2){
  Vector v;
  v.x = v1.x - v2.x;
  v.y = v1.y - v2.y;
  v.z = v1.z - v2.z;
  return v;
}

//multiplicació per un escalar
//divisió es pot fer fent multiplica(1/esc);
Vector multiplica(Vector v, double esc){
  Vector vec;
  vec.x = esc * v.x;
  vec.y = esc * v.y;
  vec.z = esc * v.z;
  return vec;
}

//producte escalar
double prod_esc(Vector v1, Vector v2){
  return v1.x*v2.x + v1.y*v2.y + v1.z*v2.z;
}

//producte vectorial
Vector prod_vec(Vector v1, Vector v2){
  Vector v;
  v.x = v1.y*v2.z - v1.z*v2.y;
  v.y = v1.z*v2.x - v1.x*v2.z;
  v.z = v1.x*v2.y - v1.y*v2.x;
  return v;
}

//modul (llargada vector)
//llargada/magnitud d'un vector
double modul(Vector v){
  return sqrt(v.x*v.x + v.y*v.y + v.z*v.z);
}

//normalitza vector per la seva llargada
Vector norma(Vector v){
  return multiplica(v, 1/modul(v));
}
